### PR TITLE
feat(extractor): Extract changelog messages from conventional commit logs

### LINKED ISSUE #15: Support generation of changelog from commits.
This isn't a small one, will leave as just reading from release_notes files for now.

### PR DESCRIPTION
**Pull request by EdgyEdgemond**
Parse commit logs for any matching https://www.conventionalcommits.org/en/v1.0.0/

Refs: #15

### CODE REVIEW

**Comment by EdgyEdgemond** [comment]
And after all these changes are merged and a release is done, would be good to switch to conventional commits.

Can do it on PR's with squash + a template for PR comment/description. And then use those as the commit message.

**Comment by EdgyEdgemond** [comment]
closes #15

File: changelog_gen/extractor.py [hunk 1]
```diff
@@ -21,37 +20,66 @@ def __init__(self: typing.Self, supported_sections: list[str], *, dry_run: bool
...
+
+                    contents = issue.read_text().strip()
+                    if section not in self.supported_sections:
+                        msg = f"Unsupported CHANGELOG section {section}"
+                        raise errors.InvalidSectionError(msg)
+
+                    sections[section][issue_ref] = {
+                        "description": contents,
+                        "breaking": breaking,
+                    }
```
Comment (myddrin) [inline changelog_gen/extractor.py:51]: 🙃 : make a method for that?

File: changelog_gen/extractor.py [hunk 2]
```diff
@@ -21,37 +20,66 @@ def __init__(self: typing.Self, supported_sections: list[str], *, dry_run: bool
...
+                    m = re.match(r"Refs: #?([\w-]+)", line)
+                    if m:
+                        issue_ref = m[1]
 
                 sections[section][issue_ref] = {
-                    "description": contents,
+                    "description": message,
                     "breaking": breaking,
+                    "scope": scope,
                 }
```
Comment (myddrin) [inline changelog_gen/extractor.py:82]: 🙃 : move to its own method that updates a given `sections` dict?

File: changelog_gen/writer.py [hunk 3]
```diff
@@ -49,19 +49,19 @@ def consume(self: typing.Self, supported_sections: dict[str, str], sections: Sec
                 continue
 
             header = supported_sections[section]
-            self.add_section(header, {k: v["description"] for k, v in sections[section].items()})
+            self.add_section(header, sections[section])
 
-    def add_section(self: typing.Self, header: str, lines: list[str]) -> None:
+    def add_section(self: typing.Self, header: str, lines: dict[str, dict]) -> None:
```
Comment (myddrin) [inline changelog_gen/writer.py:54]: 🙃 : maybe `lines` isn't the best name anymore. Should that be `section`? And maybe create a new type for it?
```
# dict [ticket ref -> info]
# info has 2 fields: `description: str` `breaking: bool
SectionData = dict[str, dict[str, str]]
# dict[section key -> section data]
SectionDict = dict[str, SectionData]
```
Maybe even make info a dataclass

File: changelog_gen/extractor.py [hunk 4]
```diff
@@ -21,37 +20,66 @@ def __init__(self: typing.Self, supported_sections: list[str], *, dry_run: bool
...
+            if m:
+                section = m[1]
+                scope = m[2]
+                breaking = m[3] is not None
+                message = m[4]
+                details = m[5] or ""
+
+                # Handle missing refs in commit message, skip link generation in writer
+                issue_ref = f"__{i}__"
+                breaking = breaking or "BREAKING CHANGE" in details
```
Comment (myddrin) [inline changelog_gen/extractor.py:72]: 🙃 : Should "BREAKING CHANGE" come from the config to support other languages?

File: changelog_gen/extractor.py (hunk 1 above)
Comment (EdgyEdgemond) [reply changelog_gen/extractor.py:51]: Yeah, this got refactored later on I believe (will link PR)

File: changelog_gen/extractor.py (hunk 2 above)
Comment (EdgyEdgemond) [reply changelog_gen/extractor.py:82]: Will check later PR, I believe this was done. Definitely the separate method

File: changelog_gen/writer.py (hunk 3 above)
Comment (EdgyEdgemond) [reply changelog_gen/writer.py:54]: This becomes `changes: dict[str, list[Change[]` in a later PR, Change being a dataclass. or something similar

File: changelog_gen/extractor.py (hunk 4 above)
Comment (EdgyEdgemond) [reply changelog_gen/extractor.py:72]: BREAKING CHANGE is a standardised footer like the `!` in type, from conventional commits

File: changelog_gen/writer.py (hunk 3 above)
Comment (EdgyEdgemond) [reply changelog_gen/writer.py:54]: ```    def add_section(self: typing.Self, header: str, changes: dict[str, Change]) -> None:```

File: changelog_gen/extractor.py (hunk 2 above)
Comment (EdgyEdgemond) [reply changelog_gen/extractor.py:82]: https://github.com/EdgyEdgemond/changelog-gen/issues/94

File: changelog_gen/extractor.py (hunk 1 above)
Comment (EdgyEdgemond) [reply changelog_gen/extractor.py:51]: https://github.com/EdgyEdgemond/changelog-gen/pull/77/files#diff-5ec09b87a087c81907fc97531029c0a529055bc916aa368dacc36cfb429fe67aR37

**Pull request merged**